### PR TITLE
[CORE] Support get native plan tree string

### PR DESCRIPTION
**Pull request by ulysses-you**
## What changes were proposed in this pull request?

This pr adds new method `substraitPlan` and `nativePlanString` in `WholeStageTransformer`, so that we can get the substrait node and native plan tree string from java interface.
Add a new config `spark.gluten.sql.cacheWholeStageTransformerContext` to check if we should cache the runtime substrait context.

For example:

```
val df = spark.sql("select split(c2, 1), c1 from t1")
df.queryExecution.executedPlan.find(_.isInstanceOf[WholeStageTransformer]).get.asInstanceOf[WholeStageTransformer].nativePlanString()

-- Project[expressions: (n1_2:ARRAY<VARCHAR>, split(try_cast "n0_1" as VARCHAR,"1")), (n1_3:BIGINT, "n0_0")] -> n1_2:ARRAY<VARCHAR>, n1_3:BIGINT
  -- TableScan[table: hive_table] -> n0_0:BIGINT, n0_1:BIGINT
```

For native plan string, this pr only implements velox backend part.

## How was this patch tested?

add test

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @zhztheplayer @PHILO-HE @zzcclp how about this ?

File: backends-velox/src/test/scala/io/glutenproject/benchmarks/NativeBenchmarkPlanGenerator.scala [hunk 1]
```diff
@@ -50,27 +51,29 @@ class NativeBenchmarkPlanGenerator extends VeloxWholeStageTransformerSuite {
   }
 
   test("Test plan json non-empty - AQE off") {
-    spark.sparkContext.setLogLevel("DEBUG")
-    withSQLConf(SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "false") {
+    withSQLConf(
+      SQLConf.ADAPTIVE_EXECUTION_ENABLED.key -> "false",
+      GlutenConfig.CACHE_WHOLE_STAGE_TRANSFORMER_CONTEXT.key -> "true") {
```
Comment (PHILO-HE) [inline backends-velox/src/test/scala/io/glutenproject/benchmarks/NativeBenchmarkPlanGenerator.scala:56]: This new config must be enabled for these tests?

File: gluten-core/src/main/scala/io/glutenproject/execution/WholeStageTransformer.scala [hunk 2]
```diff
@@ -145,9 +160,9 @@ case class WholeStageTransformer(child: SparkPlan, materializeInput: Boolean = f
       maxFields,
       printNodeId = printNodeId,
       indent)
-    if (verbose && planJson.nonEmpty) {
+    if (verbose && wholeStageTransformerContext.isDefined) {
```
Comment (PHILO-HE) [inline gluten-core/src/main/scala/io/glutenproject/execution/WholeStageTransformer.scala:163]: Can we still keep the previous logic by just checking whether it is empty?

**Review by PHILO-HE — COMMENTED** [review]
Thanks for your work! Please also document this helpful functionality.

File: backends-velox/src/test/scala/io/glutenproject/benchmarks/NativeBenchmarkPlanGenerator.scala (hunk 1 above)
Comment (ulysses-you) [reply backends-velox/src/test/scala/io/glutenproject/benchmarks/NativeBenchmarkPlanGenerator.scala:56]: yes, before, it depends on the log level and this test changes level to debug. This pr adds a new config to replace the log level. Enable this config is for restore the previous behavior.

**Review by JkSelf — COMMENTED** [review]
LGTM. Here, do we have an opportunity to call the `printPlanWithStats ` method and retrieve the metrics information for each operator, as shown below:

```
--Expand[[l_orderkey, l_partkey, l_suppkey, 0], [l_orderkey, null, l_suppkey, 1], [null, null, l_suppkey, 2]] -> l_orderkey:BIGINT, l_partkey:BIGINT, l_suppkey:BIGINT, group_id_0:BIGINT
   Output: 80357184 rows (1.47GB, 8064 batches), Cpu time: 92.50ms, Blocked wall time: 0ns, Peak memory: 256B, Memory allocations: 8064, Threads: 4
  -- TableScan[table: lineitem] -> l_suppkey:BIGINT, l_orderkey:BIGINT, l_partkey:BIGINT
     Input: 26785728 rows (752.62MB, 2688 batches), Output: 26785728 rows (752.62MB, 2688 batches), Cpu time: 2.17s, Blocked wall time: 0ns, Peak memory: 313.31MB, Memory allocations: 14270, Threads: 4, Splits: 10
```

File: gluten-core/src/main/scala/io/glutenproject/execution/WholeStageTransformer.scala (hunk 2 above)
Comment (ulysses-you) [reply gluten-core/src/main/scala/io/glutenproject/execution/WholeStageTransformer.scala:163]: It is same with the previous logic. I just change the type from `String` to `Option[String]`

**Comment by ulysses-you** [comment]
@JkSelf I plan to add native plan string in Gluten UI. For stats, I'm not sure if it is required as they are already showed at UI ?

**Comment by ulysses-you** [comment]
> Thanks for your work! Please also document this helpful functionality.

@PHILO-HE I plan to add native plan string to Gluten UI, so I can do this in a that pr if you don't mind.

**Comment by zhli1142015** [comment]
> LGTM. Here, do we have an opportunity to call the `printPlanWithStats ` method and retrieve the metrics information for each operator, as shown below:
> 
> ```
> --Expand[[l_orderkey, l_partkey, l_suppkey, 0], [l_orderkey, null, l_suppkey, 1], [null, null, l_suppkey, 2]] -> l_orderkey:BIGINT, l_partkey:BIGINT, l_suppkey:BIGINT, group_id_0:BIGINT
>    Output: 80357184 rows (1.47GB, 8064 batches), Cpu time: 92.50ms, Blocked wall time: 0ns, Peak memory: 256B, Memory allocations: 8064, Threads: 4
>   -- TableScan[table: lineitem] -> l_suppkey:BIGINT, l_orderkey:BIGINT, l_partkey:BIGINT
>      Input: 26785728 rows (752.62MB, 2688 batches), Output: 26785728 rows (752.62MB, 2688 batches), Cpu time: 2.17s, Blocked wall time: 0ns, Peak memory: 313.31MB, Memory allocations: 14270, Threads: 4, Splits: 10
> ```

I think it's better to print all operator statistics like below, as now not all metrics displayed in UI.

```
printPlanWithStats(*plan, task->taskStats(), true) shows custom operator statistics.

-> Project[expressions: (c0:INTEGER, ROW["c0"]), (p1:BIGINT, plus(ROW["c1"],1)), (p2:BIGINT, plus(ROW["c1"],ROW["u_c1"]))]
   Output: 2000 rows (154.98KB), Cpu time: 1.11ms, Blocked wall time: 0ns, Peak memory: 1.00MB, Threads: 1
      dataSourceLazyWallNanos    sum: 473.00us, count: 20, min: 11.00us, max: 96.00us
  -> HashJoin[INNER c0=u_c0]
     Output: 2000 rows (136.88KB), Cpu time: 533.54us, Blocked wall time: 223.00us, Peak memory: 2.00MB
     HashBuild: Input: 100 rows (1.31KB), Output: 0 rows (0B), Cpu time: 208.57us, Blocked wall time: 0ns, Peak memory: 1.00MB, Threads: 1
        distinctKey0       sum: 101, count: 1, min: 101, max: 101
        queuedWallNanos    sum: 125.00us, count: 1, min: 125.00us, max: 125.00us
        rangeKey0          sum: 200, count: 1, min: 200, max: 200
     HashProbe: Input: 2000 rows (118.12KB), Output: 2000 rows (136.88KB), Cpu time: 324.97us, Blocked wall time: 223.00us, Peak memory: 1.00MB, Threads: 1
        dynamicFiltersProduced    sum: 1, count: 1, min: 1, max: 1
        queuedWallNanos           sum: 24.00us, count: 1, min: 24.00us, max: 24.00us
    -> TableScan[Table: Orders]
       Input: 2000 rows (118.12KB), Raw Input: 20480 rows (72.31KB), Output: 2000 rows (118.12KB), Cpu time: 5.50ms, Blocked wall time: 10.00us, Peak memory: 1.00MB, Threads: 1, Splits: 20
          dataSourceWallNanos       sum: 2.52ms, count: 40, min: 12.00us, max: 250.00us
          dynamicFiltersAccepted    sum: 1, count: 1, min: 1, max: 1
          localReadBytes            sum: 0B, count: 1, min: 0B, max: 0B
          numLocalRead              sum: 0, count: 1, min: 0, max: 0
          numPrefetch               sum: 28, count: 1, min: 28, max: 28
          numRamRead                sum: 0, count: 1, min: 0, max: 0
          numStorageRead            sum: 140, count: 1, min: 140, max: 140
          prefetchBytes             sum: 29.51KB, count: 1, min: 29.51KB, max: 29.51KB
          queuedWallNanos           sum: 29.00us, count: 1, min: 29.00us, max: 29.00us
          ramReadBytes              sum: 0B, count: 1, min: 0B, max: 0B
          skippedSplitBytes         sum: 0B, count: 1, min: 0B, max: 0B
          skippedSplits             sum: 0, count: 1, min: 0, max: 0
          skippedStrides            sum: 0, count: 1, min: 0, max: 0
          storageReadBytes          sum: 150.25KB, count: 1, min: 150.25KB, max: 150.25KB
          totalScanTime             sum: 0ns, count: 1, min: 0ns, max: 0ns
          queryThreadIoLatency      sum: 0, count: 1, min: 0, max: 0
    -> Project[expressions: (u_c0:INTEGER, ROW["c0"]), (u_c1:BIGINT, ROW["c1"])]
       Output: 100 rows (1.31KB), Cpu time: 21.50us, Blocked wall time: 0ns, Peak memory: 0B, Threads: 1
      -> Values[100 rows in 1 vectors]
         Input: 0 rows (0B), Output: 100 rows (1.31KB), Cpu time: 12.14us, Blocked wall time: 0ns, Peak memory: 0B, Th
```

**Comment by ulysses-you** [comment]
@zhli1142015 thank you for the metics. It seems helpful, I'm fine to show total metics for the native plan string.

**Review by PHILO-HE — DISMISSED** [review]
Thanks!

**Comment by PHILO-HE** [comment]
Looks code conflict happens. Please resolve it. Thanks!

**Review by PHILO-HE — DISMISSED** [review]

**Comment by PHILO-HE** [comment]
Please check the build error. Thanks!
```
[ERROR] /opt/gluten/gluten-core/src/main/scala/io/glutenproject/execution/WholeStageTransformer.scala:118: error: value initLocalFilesNodesIndex is not a member of io.glutenproject.substrait.SubstraitContext
[ERROR]       wholeStageTransformerContext.get.substraitContext.initLocalFilesNodesIndex(0)
[ERROR]                                                         ^
[ERROR] one error found
[ERROR] exception compilation error occurred!!!
```

**Comment by zhouyuan** [comment]
@ulysses-you does this requires the query to be executed to extract the plans? 
IIRC, substrait already supports dumping the plans to some json file, would this help?
https://github.com/substrait-io/substrait-cpp/blob/main/src/substrait/textplan/converter/SaveBinary.cpp#L45
https://github.com/substrait-io/substrait-validator

thanks, -yuan

File: shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala [hunk 3]
```diff
@@ -1322,4 +1322,12 @@ object GlutenConfig {
         "if files are mutable, i.e. file content may change while file path stays the same.")
       .booleanConf
       .createWithDefault(false)
+
+  val CACHE_WHOLE_STAGE_TRANSFORMER_CONTEXT =
+    buildConf("spark.gluten.sql.cacheWholeStageTransformerContext")
+      .internal()
+      .doc("When true, `WholeStageTransformer` will cache the `WholeStageTransformerContext` " +
+        "when executing")
```
Comment (zhouyuan) [inline shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala:1330]: I'm not clear on this config, maybe need to add some extra words here?
like caching `WholeStageTransformerContext` for xxxx

Comment (ulysses-you) [reply shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala]: add some words to describle the use case.

**Comment by ulysses-you** [comment]
@zhouyuan it is not required. For dumping substrait json string, it has been supported before this pr at java side. What this pr did is:
- expose the substrait plan node object
- add a new config to cache the runtime context to make it easy to use
- expose the native plan string

File: gluten-core/src/main/scala/io/glutenproject/execution/WholeStageTransformer.scala [hunk 4]
```diff
@@ -108,13 +108,28 @@ case class WholeStageTransformer(child: SparkPlan, materializeInput: Boolean = f
...
+  }
+
+  def substraitPlanJson: String = {
+    SubstraitPlanPrinterUtil.substraitPlanToJson(substraitPlan.toProtobuf)
+  }
+
+  def nativePlanString(details: Boolean = true): String = {
+    BackendsApiManager.getTransformerApiInstance.getNativePlanString(
+      substraitPlan.toProtobuf.toByteArray,
+      details)
```
Comment (zzcclp) [inline gluten-core/src/main/scala/io/glutenproject/execution/WholeStageTransformer.scala:132]: what's the difference between using `SubstraitPlanPrinterUtil.substraitPlanToJson(substraitPlan.toProtobuf)` and `BackendsApiManager.getTransformerApiInstance.getNativePlanString`?

Comment (ulysses-you) [reply gluten-core/src/main/scala/io/glutenproject/execution/WholeStageTransformer.scala:132]: @zzcclp `getNativePlanString` is used to get the plan tree string from backend, e.g., velox and ch. (see pr description for an example).  `substraitPlanToJson` is used to show the substrait plan json.

Comment (zzcclp) [reply gluten-core/src/main/scala/io/glutenproject/execution/WholeStageTransformer.scala:132]: Got it. LGTM

**Review by yaooqinn — APPROVED** [review]

**Pull request merged**